activerecord/lib/active_record/reflection.rb

```diff
@@ -764,7 +764,16 @@ def through_reflection?
       end
 
       def klass
-        @klass ||= delegate_reflection.compute_class(class_name)
+        @klass ||= delegate_reflection.compute_class(class_name).tap do |klass|
+          if !parent_reflection.is_a?(HasAndBelongsToManyReflection) &&
+             !klass.reflections.include?(options[:through].to_s) &&
+             active_record.type_for_attribute(active_record.primary_key).type != :integer
```

Comment:
re: https://github.com/rails/rails/issues/38340 I updated this line to be:

`![:integer, :uuid].include?(active_record.type_for_attribute(active_record.primary_key).type)`

This "Works" in the sense that it allows my hacking around with the new ActiveStorage proxy feature to continue.